client/gutenberg/extensions/contact-info/phone/save.js

```diff
@@ -11,7 +11,11 @@ export function renderPhone( inputText ) {
 			if ( number.trim() === '' ) {
 				return number;
 			}
-			const just_number = number.replace( /\D/g, '' );
+			let just_number = number.replace( /\D/g, '' );
+			// Numbers starting with + shoud be part of the number.
+			if ( number.substring( 0, 1 ) === '+' ) {
+				just_number = '+' + just_number;
+			}
```

Comment:
Just another way to do it — up to you which you prefer for code-readability (if might be clearest)

```js
// Replace everything non-digit, except allow first character to be `+`
String.replace(/(?!^\+)[\D]/g, '');
```

Or the same in two passes:

```js
// First replace all non `+` and `0-9` 
// Then replace `+` except if it's in the beginning of the string
String.replace(/[^\+0-9]/g, '').replace(/(?!^)\+/g, '');
```